anoncreds/src/services/prover.rs

```diff
@@ -248,52 +254,101 @@ pub fn create_presentation(
         rev_reg_idx,
-        timestamp,
-        rev_state
+        rev_state,
+        old_rev_reg_list,
     );
 
     let RevocationRegistryDefinition::RevocationRegistryDefinitionV1(revoc_reg_def) = revoc_reg_def;
-    let RevocationRegistryDelta::RevocationRegistryDeltaV1(rev_reg_delta) = rev_reg_delta;
-
-    let witness = match rev_state {
-        None => Witness::new(
-            rev_reg_idx,
-            revoc_reg_def.value.max_cred_num,
-            revoc_reg_def.value.issuance_type.to_bool(),
-            &rev_reg_delta.value,
-            &tails_reader,
-        )?,
-        Some(source_rev_state) => {
+    let mut issued = HashSet::<u32>::new();
+    let mut revoked = HashSet::<u32>::new();
+    let witness =
+        if let (Some(source_rev_state), Some(source_rev_list)) = (rev_state, old_rev_reg_list) {
+            _create_index_deltas(
+                rev_reg_list.state_owned().bitxor(source_rev_list.state()),
+                rev_reg_list.state(),
+                &mut issued,
+                &mut revoked,
+            );
+
+            let rev_reg_delta = RevocationRegistryDelta::from_parts(
+                Some(&source_rev_list.into()),
+                &rev_reg_list.into(),
+                &issued,
+                &revoked,
+            );
             let mut witness = source_rev_state.witness.clone();
             witness.update(
                 rev_reg_idx,
                 revoc_reg_def.value.max_cred_num,
-                &rev_reg_delta.value,
+                &rev_reg_delta,
                 &tails_reader,
             )?;
             witness
-        }
-    };
+        } else {
+            let list_size = usize::try_from(revoc_reg_def.value.max_cred_num)
+                .map_err(|e| Error::from_msg(crate::ErrorKind::InvalidState, e.to_string()))?;
+            let list = match revoc_reg_def.value.issuance_type {
+                // All cred are revoked by default
+                IssuanceType::ISSUANCE_ON_DEMAND => {
+                    bitvec![1; list_size]
+                }
+                IssuanceType::ISSUANCE_BY_DEFAULT => {
+                    bitvec![0; list_size]
+                }
+            };
+            _create_index_deltas(
+                rev_reg_list.state_owned().bitxor(list),
+                rev_reg_list.state(),
+                &mut issued,
+                &mut revoked,
+            );
+            let rev_reg_delta =
+                RevocationRegistryDelta::from_parts(None, &rev_reg_list.into(), &issued, &revoked);
+            Witness::new(
+                rev_reg_idx,
+                revoc_reg_def.value.max_cred_num,
+                revoc_reg_def.value.issuance_type.to_bool(),
+                &rev_reg_delta,
+                &tails_reader,
+            )?
+        };
 
     Ok(CredentialRevocationState {
         witness,
-        rev_reg: CryptoRevocationRegistry::from(rev_reg_delta.value.clone()),
-        timestamp,
+        rev_reg: rev_reg_list.into(),
```

Comment:
Could we not change the type of the state to contain  just the `rev_reg_list`? Now we have the timestamp in there twice.

Reply:
The `rev_reg_list` is of type `RevocationList` which fields such as the `revocation_list` which is not needed.
`CredentialRevocationState` does not contain `RevocationList` so the timestamp is only there once.